core/src/test/scala/integration/kafka/server/DynamicBrokerReconfigurationTest.scala

```diff
@@ -373,10 +371,101 @@ class DynamicBrokerReconfigurationTest extends ZooKeeperTestHarness with SaslSet
+    newProps.setProperty("some.prop", "some.value")
+    reconfigureServers(newProps, perBrokerConfig = false, (TestMetricsReporter.PollingIntervalProp, "100"))
+    reporters.foreach(_.verifyState(reconfigureCount = 0, deleteCount = 0, pollingInterval = 100))
+
+    // Update of custom config of metrics reporter should reconfigure reporter
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "1000")
+    reconfigureServers(newProps, perBrokerConfig = false, (TestMetricsReporter.PollingIntervalProp, "1000"))
+    reporters.foreach(_.verifyState(reconfigureCount = 1, deleteCount = 0, pollingInterval = 1000))
+
+    // Verify removal of metrics reporter
+    configureMetricsReporters(Seq.empty[Class[_]], newProps)
+    reporters.foreach(_.verifyState(reconfigureCount = 1, deleteCount = 1, pollingInterval = 1000))
+    TestMetricsReporter.testReporters.clear()
+
+    // Verify recreation of metrics reporter
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "2000")
+    configureMetricsReporters(Seq(classOf[TestMetricsReporter]), newProps)
+    val newReporters = waitForReporters()
+    newReporters.foreach(_.verifyState(reconfigureCount = 0, deleteCount = 0, pollingInterval = 2000))
+
+    // Verify that validation failure of metrics reporter fails reconfiguration and leaves config unchanged
+    newProps.put(KafkaConfig.MetricReporterClassesProp, "unknownMetricsReporter")
+    reconfigureServers(newProps, perBrokerConfig = false, (TestMetricsReporter.PollingIntervalProp, "2000"), expectFailure = true)
+    servers.foreach { server =>
+      assertEquals(classOf[TestMetricsReporter].getName, server.config.originals.get(KafkaConfig.MetricReporterClassesProp))
+    }
+    newReporters.foreach(_.verifyState(reconfigureCount = 1, deleteCount = 0, pollingInterval = 2000))
+
+    // Verify that validation failure of custom config fails reconfiguration and leaves config unchanged
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "invalid")
+    reconfigureServers(newProps, perBrokerConfig = false, (TestMetricsReporter.PollingIntervalProp, "2000"), expectFailure = true)
+    newReporters.foreach(_.verifyState(reconfigureCount = 1, deleteCount = 0, pollingInterval = 2000))
+
+    // Delete reporters
+    configureMetricsReporters(Seq.empty[Class[_]], newProps)
+    TestMetricsReporter.testReporters.clear()
+
+    // Verify that even though metrics reporters can be defined at default cluster level for consistent
+    // configuration across brokers, they can also be defined at per-broker level for testing
+    newProps.put(KafkaConfig.MetricReporterClassesProp, classOf[TestMetricsReporter].getName)
+    newProps.put(TestMetricsReporter.PollingIntervalProp, "4000")
+    alterConfigsOnServer(servers.head, newProps)
+    TestUtils.waitUntilTrue(() => !TestMetricsReporter.testReporters.isEmpty, "Metrics reporter not created")
+    val perBrokerReporter = TestMetricsReporter.testReporters.peek()
```

Comment:
Is it worth asserting that the size of the collection is 1?

Reply:
I have switched this to use the updated `waitForReporters` method which checks the size.